Fix a data race in kubelet image manager

roycaihw opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Let image manager return a copy of image list. Today it [sorts and returns](https://github.com/kubernetes/kubernetes/blob/7c7ce47c018415b09e430c62216764b1b952e19f/pkg/kubelet/images/image_gc_manager.go#L130-L131) the internal image list cache to caller directly, which introduces a data race. Due to the sort being in-place, we've observed the race causing static pod workers to panic during [syncPod](https://github.com/kubernetes/kubernetes/blob/7c7ce47c018415b09e430c62216764b1b952e19f/pkg/kubelet/kubelet.go#L1657). However, given the way kubelet [recovers](https://github.com/kubernetes/kubernetes/blob/7c7ce47c018415b09e430c62216764b1b952e19f/pkg/kubelet/pod_workers.go#L220-L223) from goroutine panics, kubelet would keep running and the pod worker silently stopped working-- all future work dispatched for this pod got ignored. 

git log shows the race was introduced in [1.10](https://github.com/kubernetes/kubernetes/pull/60159). We should cherrypick this fix to all the supported minor releases.

Sample panic log from 1.13:
```
runtime.go:73] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 755 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x..., 0x...)
	...k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:69 +0x...
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x..., 0x..., 0x...)
	...k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51 +0x...
panic(0x..., 0x...)
	/usr/local/go/src/runtime/panic.go:522 +0x...
k8s.io/kubernetes/pkg/kubelet/nodestatus.Images.func1(0x..., 0x..., 0x...)
	...k8s.io/kubernetes/pkg/kubelet/nodestatus/setters.go:413 +0x...
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).setNodeStatus(0x..., 0x...)
	...k8s.io/kubernetes/pkg/kubelet/kubelet_node_status.go:480 +0x...
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).initialNode(0x..., 0x..., 0x..., 0x...)
	...k8s.io/kubernetes/pkg/kubelet/kubelet_node_status.go:348 +0x...
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).GetNode(0x..., 0x..., 0x..., 0x...)
	...k8s.io/kubernetes/pkg/kubelet/kubelet_getters.go:220 +0x...
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncPod(0x..., 0x..., 0x..., 0x..., 0x..., 0x..., 0x..., 0x...)
	...k8s.io/kubernetes/pkg/kubelet/kubelet.go:1666 +0x...
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).managePodLoop.func1(0x..., 0x..., 0x..., 0x..., 0x...)
	...k8s.io/kubernetes/pkg/kubelet/pod_workers.go:174 +0x...
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).managePodLoop(0x..., 0x...)
	...k8s.io/kubernetes/pkg/kubelet/pod_workers.go:183 +0x...
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod.func1(0x..., 0x...)
	...k8s.io/kubernetes/pkg/kubelet/pod_workers.go:221 +0x...
created by k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod
	...k8s.io/kubernetes/pkg/kubelet/pod_workers.go:219 +0x...
```

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #60443

**Special notes for your reviewer**:
I'm going to add a proposal to SIG node about catching/mitigating this kind of programming error in future. In this case the issue would have been mitigated if we either 1) actually crashed kubelet and let it restart or 2) created a new goroutine when the old one panicked. Restarting kubelet is preferred IMO because it guards a wide range of programming error, and it's usually safe to restart kubelet as long as it's not too frequent.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a data race in kubelet image manager that can cause static pod workers to silently stop working.
```

/cc @Random-Liu @cheftako 
/priority critical-urgent
/sig node

roycaihw commented:
I'm not sure what additional approval we need given it's code freeze. I will prepare the cherrypicks once it's merged in master

Random-Liu commented on pkg/kubelet/images/image_gc_manager_test.go:
```diff
@@ -548,14 +550,19 @@ func TestValidateImageGCPolicy(t *testing.T) {
 	}
 }
 
+//
```
?

roycaihw replied:
ah, I was going to add a test description but decided to not at some point. I updated the test comments and the variable name to `expect` to make things more clear.

Random-Liu commented on pkg/kubelet/images/image_gc_manager_test.go:
```diff
@@ -548,14 +550,19 @@ func TestValidateImageGCPolicy(t *testing.T) {
+	testListCopy := append([]container.Image{}, testList...)
+	// Expect cache.get to sort the image list
+	sort.Sort(sliceutils.ByImageSize(testListCopy))
 	cache.set(testList)
```
We should also test that `testList` is not changed after `get`.

roycaihw replied:
It is changed. That's why the test didn't catch the bug before. Please see the comment I added for `imageCache.set()`

It brings another question: should `imageCache.set()` make a copy of the given list? E.g.
```
func (i *imageCache) set(images []container.Image) {
	i.Lock()
	defer i.Unlock()
	imagesCopy := append([]container.Image{}, images...)
	i.images = imagesCopy
}
```

this way the caller of `imageCache.set()` can safely reference `images` afterwards. I think it's a reasonable change but not as critical. Currently [this goroutine](https://github.com/kubernetes/kubernetes/blob/1836f952600c08ac3bdc21e38c72bd5d28135ed5/pkg/kubelet/images/image_gc_manager.go#L194) is the only caller of `imageCache.set()` and it doesn't reference `images` afterwards, so the change doesn't matter. Since we are going to cherrypick this fix, I prefer keeping the minimum change needed. That's why I didn't change `imageCache.set()` behavior here.

roycaihw replied:
I moved the sort logic outside of the image cache and tested that `testList` is not changed after `get`

Random-Liu commented:
This is a critical bug fix! Mark 1.18.

roycaihw commented:
/test pull-kubernetes-e2e-kind

roycaihw commented:
/test pull-kubernetes-integration

liggitt commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -113,22 +113,23 @@ type imageCache struct {
-	sort.Sort(sliceutils.ByImageSize(i.images))
-	return i.images
+	// Make a copy of image list to avoid data race
+	images := append([]container.Image{}, i.images...)
```
How frequently is this read? Should we sort on set and return the list without allocating?

roycaihw replied:
That's a good point! It's invoked in different places so it's hard to tell from source. I searched [this](https://github.com/kubernetes/kubernetes/blob/0bb125e731c70c0a4a6a022694629d8fef868e8d/pkg/kubelet/kubelet_node_status.go#L294-L297) and [this](https://github.com/kubernetes/kubernetes/blob/0bb125e731c70c0a4a6a022694629d8fef868e8d/pkg/kubelet/kubelet_node_status.go#L342-L363) log messages in a master node and found the [read](https://github.com/kubernetes/kubernetes/blob/0bb125e731c70c0a4a6a022694629d8fef868e8d/pkg/kubelet/kubelet_node_status.go#L369) is invoked every few (<5) seconds (see the log below). Compared to the write getting invoked [every 30s](https://github.com/kubernetes/kubernetes/blob/0bb125e731c70c0a4a6a022694629d8fef868e8d/pkg/kubelet/images/image_gc_manager.go#L187-L196), I agree it makes sense to sort on set. 

Regarding whether we should skip allocating on return, I'm not sure if it's safe to do. Currently we [truncate the length of the returned list](https://github.com/kubernetes/kubernetes/blob/0bb125e731c70c0a4a6a022694629d8fef868e8d/pkg/kubelet/nodestatus/setters.go#L445), which has no effect on the original list https://play.golang.org/p/fc6JmArJtfW. But I still feel it's safer if we copy on return, to prevent the original list getting mutated in future. 

```
I0309 21:40:27.927125    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:35.971794    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:39.926950    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:43.582650    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:44.926879    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:45.926867    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:46.104704    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:52.927557    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:53.927034    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:56.239125    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:40:56.926903    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:06.404693    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:08.926829    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:11.926992    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:14.926864    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:16.541316    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:19.929088    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:20.926684    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:23.926947    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:23.927801    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:26.720531    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:28.926814    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:32.926843    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:32.927004    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:34.926880    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:36.884823    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:37.926909    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:38.926790    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:39.927096    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:43.582911    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:43.927033    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:47.013040    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:54.926907    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:55.927053    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:56.927574    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:41:57.161981    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:02.927006    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:04.926781    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:05.927033    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:07.335855    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:13.927023    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:17.502770    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:19.927029    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:21.926896    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:27.638850    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
I0309 21:42:32.926875    1438 kubelet_node_status.go:279] Setting node annotation to enable volume controller attach/detach 
```

roycaihw commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -191,6 +190,10 @@ func (im *realImageGCManager) Start() {
+			// The image list needs to be sorted when it gets read and used in
+			// setNodeStatusImages. We sort the list on write instead of on read,
+			// because the image cache is more often read than written
+			sort.Sort(sliceutils.ByImageSize(images))
```
I was almost going to add another copy before this line... Perhaps it's better to put the copy and the sort inside `set()`, to make the code more contained? @liggitt WDYT?

liggitt replied:
It looks like that's what used to happen:

https://github.com/kubernetes/kubernetes/blob/d5f1d5286fab4c730a5fbf8c38a07fab6ad2dba6/pkg/kubelet/images/image_gc_manager.go#L190-L194

copying/sorting on write makes sense to me if that statement about being read more frequently than written is true. Do we know why that was changed?

liggitt replied:
moving copy and sort into set makes sense to me

roycaihw committed:
let image cache do sort on write instead of on read to avoid data

race and improve efficienty

roycaihw commented:
@liggitt @Random-Liu Thanks for the review! I've updated the change. PTAL

Random-Liu commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -113,22 +113,25 @@ type imageCache struct {
 func (i *imageCache) set(images []container.Image) {
 	i.Lock()
 	defer i.Unlock()
+	// make a copy to avoid potential data race
```
I think we can't avoid the data race, because the image list is passed in, and the caller may concurrently do some operation on it.

However, I do agree that copying it here is better. :)

roycaihw replied:
Ah you're right. I updated the comment to document that the caller should not reference the list again (similar to the comments for [client-go cache](https://github.com/kubernetes/kubernetes/blob/f8ff8f44206ff4dd9b58386d96462b01a3d79fb1/staging/src/k8s.io/client-go/tools/cache/store.go#L231-L233) and [fifo queue](https://github.com/kubernetes/kubernetes/blob/f8ff8f44206ff4dd9b58386d96462b01a3d79fb1/staging/src/k8s.io/client-go/tools/cache/fifo.go#L313-L315))
```
 // set sorts the input list and updates image cache.
+// 'i' takes ownership of the list, you should not reference the list again
+// after calling this function.
 func (i *imageCache) set(images []container.Image) {
        i.Lock()
        defer i.Unlock()
-       // make a copy to avoid potential data race
+       // Make a copy of the input list. This doesn't completely avoid
+       // potential data race. The caller should not reference the list again
+       // after calling this function.
```

Random-Liu reviewed (commented):
/lgtm
/approve

liggitt commented on pkg/kubelet/images/image_gc_manager.go:
```diff
@@ -113,22 +113,29 @@ type imageCache struct {
 	defer i.Unlock()
+	// Make a copy of the input list. This doesn't completely avoid
+	// potential data race. The caller should not reference the list again
+	// after calling this function.
```
I think this comment is unnecessarily alarming... the set() method is private, and the only place we call it does not hold onto the list and change it further. The godoc on the method clarifying ownership is probably sufficient.

roycaihw replied:
thanks. Removed the comment

liggitt commented:
Traced all setters/getters of the function, and none are modifying the slice. There is one place that is modifying slice members of the image items in the list:

https://github.com/kubernetes/kubernetes/blob/c4a7d3cf163484f4a84abb8fea0c55a550bd3b45/pkg/kubelet/nodestatus/setters.go#L449

That should change to accumulate names in a new `[]string` var

roycaihw commented:
> That should change to accumulate names in a new []string var

Fixed that line. 

For the record, I talked with @liggitt offline: making copies for the image list could be expensive given that it's done every few seconds, and I'm not sure how big these lists are (seems that it can be [greater than 50](https://github.com/kubernetes/kubernetes/blob/789dc873f6816cc2b9b39e77a9b94f478d3a3134/cmd/kubelet/app/options/options.go#L221)). Since these functions are internal, for efficiency we decided to not make copies, and document these functions to make it clear that the callers should not reference/modify the input/output to introduce race. I also removed the unit test because the get() isn't providing "copy" guarantee now (and the test didn't ensure that)-- we depend on the callers to behave well and avoid race.

roycaihw commented:
/retest

liggitt commented:
/lgtm
/approve
/retest

liggitt commented:
see also https://github.com/kubernetes/kubernetes/issues/88779 for issues with the kubelet not exiting on panic

liggitt commented:
/retest

roycaihw commented:
/retest
